# Movie Recommendation content Based System

In [2]:
import pandas as pd
import numpy as np
from warnings import filterwarnings
filterwarnings('ignore')

In [3]:
credits= pd.read_csv('tmdb_5000_credits.csv')
movies= pd.read_csv('tmdb_5000_movies.csv')
credits.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

In [4]:
movies.shape,credits.shape

((4803, 20), (4803, 4))

In [5]:
movies.head()

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4  [{"id": 818, "name": "based on novel"}, {"id":...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   
3  Following the death of District Attorney Harve...  112.312950   
4  John Carter is a war-weary, former military ca...   43.926995   

                                production_companies  \
0  [{"name": "Ingenious Film Partners", "id": 289...   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...   
3  [{"name": "Legendary Pictures", "id": 923}, {"...   
4        [{"name": "Walt Disney Pictures", "id": 2}]   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   
1  [{"iso_3166_1": "US", "name": "United States o...   2007-05-19   961000000   
2  [{"iso_3166_1": "GB", "name": "United Kingdom"...   2015-10-26   880674609   
3  [{"iso_3166_1": "US", "name": "United States o...   2012-07-16  1084939099   
4  [{"iso_3166_1": "US", "name": "United States o...   2012-03-07   284139100   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1    169.0           [{"iso_639_1": "en", "name": "English"}]  Released   
2    148.0  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3    165.0           [{"iso_639_1": "en", "name": "English"}]  Released   
4    132.0           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                      title  vote_average  vote_count  
0                                    Avatar           7.2       11800  
1  Pirates of the Caribbean: At World's End           6.9        4500  
2                                   Spectre           6.3        4466  
3                     The Dark Knight Rises           7.6        9106  
4                               John Carter           6.1        2124

In [6]:
movie_df=movies.merge(credits,on='title')
movie_df.shape

(4809, 23)

In [7]:
movie_df.columns

Index(['budget', 'genres', 'homepage', 'id', 'keywords', 'original_language',
       'original_title', 'overview', 'popularity', 'production_companies',
       'production_countries', 'release_date', 'revenue', 'runtime',
       'spoken_languages', 'status', 'tagline', 'title', 'vote_average',
       'vote_count', 'movie_id', 'cast', 'crew'],
      dtype='object')

In [8]:
# only keeping selected columns
# movie_id,title,cast,overview,spoken_languages,'genres',keywords
old_df=movie_df
movie_df=movie_df[['movie_id','title','cast','crew','overview','spoken_languages','genres','keywords']]

In [9]:
movie_df.isnull().sum()

movie_id            0
title               0
cast                0
crew                0
overview            3
spoken_languages    0
genres              0
keywords            0
dtype: int64

In [10]:
movie_df.dropna(axis=0,inplace=True)

In [11]:
movie_df.drop_duplicates(inplace=True)

### feature engineering

genres

In [12]:
movie_df['genres'].values[0]

'[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]'

In [13]:
# get genres name
import ast
def convert(obj):
    L=[]
    for i in ast.literal_eval(obj):
        L.append(i['name'])
    return L

string='[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]'
convert(string)

['Action', 'Adventure', 'Fantasy', 'Science Fiction']

In [14]:
movie_df['genres']=movie_df['genres'].apply(convert)
movie_df['genres'].values[:3]

array([list(['Action', 'Adventure', 'Fantasy', 'Science Fiction']),
       list(['Adventure', 'Fantasy', 'Action']),
       list(['Action', 'Adventure', 'Crime'])], dtype=object)

Keywords

In [15]:
movie_df['keywords'].values[:3]

array(['[{"id": 1463, "name": "culture clash"}, {"id": 2964, "name": "future"}, {"id": 3386, "name": "space war"}, {"id": 3388, "name": "space colony"}, {"id": 3679, "name": "society"}, {"id": 3801, "name": "space travel"}, {"id": 9685, "name": "futuristic"}, {"id": 9840, "name": "romance"}, {"id": 9882, "name": "space"}, {"id": 9951, "name": "alien"}, {"id": 10148, "name": "tribe"}, {"id": 10158, "name": "alien planet"}, {"id": 10987, "name": "cgi"}, {"id": 11399, "name": "marine"}, {"id": 13065, "name": "soldier"}, {"id": 14643, "name": "battle"}, {"id": 14720, "name": "love affair"}, {"id": 165431, "name": "anti war"}, {"id": 193554, "name": "power relations"}, {"id": 206690, "name": "mind and soul"}, {"id": 209714, "name": "3d"}]',
       '[{"id": 270, "name": "ocean"}, {"id": 726, "name": "drug abuse"}, {"id": 911, "name": "exotic island"}, {"id": 1319, "name": "east india trading company"}, {"id": 2038, "name": "love of one\'s life"}, {"id": 2052, "name": "traitor"}, {"id": 2580,

In [16]:
movie_df['keywords']=movie_df['keywords'].apply(convert)

In [17]:
movie_df['spoken_languages'].apply(convert).value_counts()

[English]                    3174
[English, Español]            127
[English, Français]           115
[]                             85
[Español, English]             54
                             ... 
[English, 普通话, Pусский]         1
[Tiếng Việt, English]           1
[Latin, English]                1
[Latin, English, Español]       1
[Polski, English]               1
Name: spoken_languages, Length: 530, dtype: int64

In [18]:
# Drop spoken_languages
movie_df.drop(['spoken_languages'],axis=1,inplace=True)

In [19]:
movie_df.head(1)

movie_id   title                                               cast  \
0     19995  Avatar  [{"cast_id": 242, "character": "Jake Sully", "...   

                                                crew  \
0  [{"credit_id": "52fe48009251416c750aca23", "de...   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   

                                            keywords  
0  [culture clash, future, space war, space colon...

<b>Casts</b> : Extract Top 3 Charecter Actors and Actresses

In [20]:
movie_df['cast'][1]

'[{"cast_id": 4, "character": "Captain Jack Sparrow", "credit_id": "52fe4232c3a36847f800b50d", "gender": 2, "id": 85, "name": "Johnny Depp", "order": 0}, {"cast_id": 5, "character": "Will Turner", "credit_id": "52fe4232c3a36847f800b511", "gender": 2, "id": 114, "name": "Orlando Bloom", "order": 1}, {"cast_id": 6, "character": "Elizabeth Swann", "credit_id": "52fe4232c3a36847f800b515", "gender": 1, "id": 116, "name": "Keira Knightley", "order": 2}, {"cast_id": 12, "character": "William \\"Bootstrap Bill\\" Turner", "credit_id": "52fe4232c3a36847f800b52d", "gender": 2, "id": 1640, "name": "Stellan Skarsg\\u00e5rd", "order": 3}, {"cast_id": 10, "character": "Captain Sao Feng", "credit_id": "52fe4232c3a36847f800b525", "gender": 2, "id": 1619, "name": "Chow Yun-fat", "order": 4}, {"cast_id": 9, "character": "Captain Davy Jones", "credit_id": "52fe4232c3a36847f800b521", "gender": 2, "id": 2440, "name": "Bill Nighy", "order": 5}, {"cast_id": 7, "character": "Captain Hector Barbossa", "credit_

In [21]:
import ast
def get_3_cast(obj):
    L=[]
    for i in ast.literal_eval(obj)[:3]:
        L.append(i['name'])
    return L
movie_df['cast'].apply(get_3_cast)

0        [Sam Worthington, Zoe Saldana, Sigourney Weaver]
1           [Johnny Depp, Orlando Bloom, Keira Knightley]
2            [Daniel Craig, Christoph Waltz, Léa Seydoux]
3            [Christian Bale, Michael Caine, Gary Oldman]
4          [Taylor Kitsch, Lynn Collins, Samantha Morton]
                              ...                        
4804    [Carlos Gallardo, Jaime de Hoyos, Peter Marqua...
4805         [Edward Burns, Kerry Bishé, Marsha Dietlein]
4806           [Eric Mabius, Kristin Booth, Crystal Lowe]
4807            [Daniel Henney, Eliza Coupe, Bill Paxton]
4808    [Drew Barrymore, Brian Herzlinger, Corey Feldman]
Name: cast, Length: 4806, dtype: object

In [22]:
movie_df['cast']=movie_df['cast'].apply(get_3_cast)

<b>Crew</b> : Extract Directer Name.

In [23]:
movie_df['crew'][0]

'[{"credit_id": "52fe48009251416c750aca23", "department": "Editing", "gender": 0, "id": 1721, "job": "Editor", "name": "Stephen E. Rivkin"}, {"credit_id": "539c47ecc3a36810e3001f87", "department": "Art", "gender": 2, "id": 496, "job": "Production Design", "name": "Rick Carter"}, {"credit_id": "54491c89c3a3680fb4001cf7", "department": "Sound", "gender": 0, "id": 900, "job": "Sound Designer", "name": "Christopher Boyes"}, {"credit_id": "54491cb70e0a267480001bd0", "department": "Sound", "gender": 0, "id": 900, "job": "Supervising Sound Editor", "name": "Christopher Boyes"}, {"credit_id": "539c4a4cc3a36810c9002101", "department": "Production", "gender": 1, "id": 1262, "job": "Casting", "name": "Mali Finn"}, {"credit_id": "5544ee3b925141499f0008fc", "department": "Sound", "gender": 2, "id": 1729, "job": "Original Music Composer", "name": "James Horner"}, {"credit_id": "52fe48009251416c750ac9c3", "department": "Directing", "gender": 2, "id": 2710, "job": "Director", "name": "James Cameron"},

In [24]:
def get_dir(obj):
    L=[]
    for i in ast.literal_eval(obj):
        if i['job']=='Director':
            L.append(i['name'])
            break
    return L
movie_df['crew'].apply(get_dir)

0           [James Cameron]
1          [Gore Verbinski]
2              [Sam Mendes]
3       [Christopher Nolan]
4          [Andrew Stanton]
               ...         
4804     [Robert Rodriguez]
4805         [Edward Burns]
4806          [Scott Smith]
4807          [Daniel Hsia]
4808     [Brian Herzlinger]
Name: crew, Length: 4806, dtype: object

In [25]:
movie_df['crew']=movie_df['crew'].apply(get_dir)

Overview

In [26]:
movie_df['overview'][0]

'In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization.'

In [27]:
movie_df['overview'].apply(lambda x:x.split())

0       [In, the, 22nd, century,, a, paraplegic, Marin...
1       [Captain, Barbossa,, long, believed, to, be, d...
2       [A, cryptic, message, from, Bond’s, past, send...
3       [Following, the, death, of, District, Attorney...
4       [John, Carter, is, a, war-weary,, former, mili...
                              ...                        
4804    [El, Mariachi, just, wants, to, play, his, gui...
4805    [A, newlywed, couple's, honeymoon, is, upended...
4806    ["Signed,, Sealed,, Delivered", introduces, a,...
4807    [When, ambitious, New, York, attorney, Sam, is...
4808    [Ever, since, the, second, grade, when, he, fi...
Name: overview, Length: 4806, dtype: object

In [28]:
movie_df['overview']=movie_df['overview'].apply(lambda x:x.split())
movie_df.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                               cast                 crew  \
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]      [James Cameron]   
1     [Johnny Depp, Orlando Bloom, Keira Knightley]     [Gore Verbinski]   
2      [Daniel Craig, Christoph Waltz, Léa Seydoux]         [Sam Mendes]   
3      [Christian Bale, Michael Caine, Gary Oldman]  [Christopher Nolan]   
4    [Taylor Kitsch, Lynn Collins, Samantha Morton]     [Andrew Stanton]   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   
4  [John, Carter, is, a, war-weary,, former, mili...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                            keywords  
0  [culture clash, future, space war, space colon...  
1  [ocean, drug abuse, exotic island, east india ...  
2  [spy, based on novel, secret agent, sequel, mi...  
3  [dc comics, crime fighter, terrorist, secret i...  
4  [based on novel, mars, medallion, space travel...

In [30]:
# replacing the gap between words Science Fiction would became ScienceFiction
movie_df['genres'].apply(lambda x: [i.replace(' ','') for i in x])

0       [Action, Adventure, Fantasy, ScienceFiction]
1                       [Adventure, Fantasy, Action]
2                         [Action, Adventure, Crime]
3                   [Action, Crime, Drama, Thriller]
4                [Action, Adventure, ScienceFiction]
                            ...                     
4804                       [Action, Crime, Thriller]
4805                               [Comedy, Romance]
4806               [Comedy, Drama, Romance, TVMovie]
4807                                              []
4808                                   [Documentary]
Name: genres, Length: 4806, dtype: object

In [31]:
movie_df['genres']=movie_df['genres'].apply(lambda x: [i.replace(' ','') for i in x])
movie_df['crew']=movie_df['crew'].apply(lambda x: [i.replace(' ','') for i in x])
movie_df['cast']=movie_df['cast'].apply(lambda x: [i.replace(' ','') for i in x])

In [32]:
movie_df.head(3)

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   

                                            cast             crew  \
0  [SamWorthington, ZoeSaldana, SigourneyWeaver]   [JamesCameron]   
1     [JohnnyDepp, OrlandoBloom, KeiraKnightley]  [GoreVerbinski]   
2      [DanielCraig, ChristophWaltz, LéaSeydoux]      [SamMendes]   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   

                                         genres  \
0  [Action, Adventure, Fantasy, ScienceFiction]   
1                  [Adventure, Fantasy, Action]   
2                    [Action, Adventure, Crime]   

                                            keywords  
0  [culture clash, future, space war, space colon...  
1  [ocean, drug abuse, exotic island, east india ...  
2  [spy, based on novel, secret agent, sequel, mi...

In [30]:
# movie_df['keywords']=movie_df['keywords'].apply(convert)

Create tags that can be used for comparing similarity between two movies

In [33]:
movie_df['tags']=movie_df['overview']+movie_df['genres']+movie_df['keywords']+movie_df['cast']+movie_df['crew']
movie_df.head(2)

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   

                                            cast             crew  \
0  [SamWorthington, ZoeSaldana, SigourneyWeaver]   [JamesCameron]   
1     [JohnnyDepp, OrlandoBloom, KeiraKnightley]  [GoreVerbinski]   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   

                                         genres  \
0  [Action, Adventure, Fantasy, ScienceFiction]   
1                  [Adventure, Fantasy, Action]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   

                                                tags  
0  [In, the, 22nd, century,, a, paraplegic, Marin...  
1  [Captain, Barbossa,, long, believed, to, be, d...

In [34]:
# new_df
new_df=movie_df[['movie_id','title','tags']]
new_df.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                tags  
0  [In, the, 22nd, century,, a, paraplegic, Marin...  
1  [Captain, Barbossa,, long, believed, to, be, d...  
2  [A, cryptic, message, from, Bond’s, past, send...  
3  [Following, the, death, of, District, Attorney...  
4  [John, Carter, is, a, war-weary,, former, mili...

In [35]:
new_df['tags']=new_df['tags'].apply(lambda x: ' '.join(x))

In [36]:
new_df['tags'][0]

'In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization. Action Adventure Fantasy ScienceFiction culture clash future space war space colony society space travel futuristic romance space alien tribe alien planet cgi marine soldier battle love affair anti war power relations mind and soul 3d SamWorthington ZoeSaldana SigourneyWeaver JamesCameron'

In [37]:
new_df.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                tags  
0  In the 22nd century, a paraplegic Marine is di...  
1  Captain Barbossa, long believed to be dead, ha...  
2  A cryptic message from Bond’s past sends him o...  
3  Following the death of District Attorney Harve...  
4  John Carter is a war-weary, former military ca...

Use Lemmatization to reduce a word into its root_word

In [38]:
from nltk.stem.wordnet import WordNetLemmatizer

wl = WordNetLemmatizer()

In [41]:
wl.lemmatize('amazing',pos='v'),wl.lemmatize('fulfilling',pos='v')

('amaze', 'fulfil')

In [42]:
def lemmatize(text):
    y = []
    for i in text.split():
        y.append(wl.lemmatize(i,pos='v'))
    return ' '.join(y)
new_df['tags'].apply(lemmatize)

0       In the 22nd century, a paraplegic Marine be di...
1       Captain Barbossa, long believe to be dead, hav...
2       A cryptic message from Bond’s past send him on...
3       Following the death of District Attorney Harve...
4       John Carter be a war-weary, former military ca...
                              ...                        
4804    El Mariachi just want to play his guitar and c...
4805    A newlywed couple's honeymoon be upend by the ...
4806    "Signed, Sealed, Delivered" introduce a dedica...
4807    When ambitious New York attorney Sam be send t...
4808    Ever since the second grade when he first saw ...
Name: tags, Length: 4806, dtype: object

In [43]:
new_df['tags']=new_df['tags'].apply(lemmatize)

In [44]:
new_df.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                tags  
0  In the 22nd century, a paraplegic Marine be di...  
1  Captain Barbossa, long believe to be dead, hav...  
2  A cryptic message from Bond’s past send him on...  
3  Following the death of District Attorney Harve...  
4  John Carter be a war-weary, former military ca...

Use CountVectorizer to extract Top 6000 Words frequency wise and turn them into vectors.
<Br/>
So Final Array would be 4806 X 6000 size, 4806 are rows and 6000 are vectors of words

In [45]:
from sklearn.feature_extraction.text import CountVectorizer

cv = CountVectorizer(max_features=6000,stop_words='english')

In [46]:
cv.fit_transform(new_df['tags']).toarray().shape

(4806, 6000)

In [47]:
vectors = cv.fit_transform(new_df['tags']).toarray()
vectors

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

In [48]:
len(cv.get_feature_names_out())

6000

In [49]:
cv.get_feature_names_out()

array(['000', '007', '10', ..., 'zoo', 'zooeydeschanel', 'zookeeper'],
      dtype=object)

Use cosine similarity to get the similarity between two vectors

In [50]:
from sklearn.metrics.pairwise import cosine_similarity

In [51]:
cosine_similarity(vectors).shape

(4806, 4806)

Our Matrix Represents similarity in the following way
```
[
    similarity_with_first_vector,similarity_with_second_vector,similarity_with_third_vector,
....,similarity_with_nth_vector
]
```

In [52]:
similarity=cosine_similarity(vectors)

In [53]:
similarity[0],similarity[0].shape

(array([1.        , 0.0727393 , 0.05157106, ..., 0.03333333, 0.01840525,
        0.        ]),
 (4806,))

In [54]:
new_df[new_df['title']=='Avatar']

movie_id   title                                               tags
0     19995  Avatar  In the 22nd century, a paraplegic Marine be di...

In [55]:
list(enumerate(similarity[0]))

[(0, 1.0000000000000002),
 (1, 0.0727392967453308),
 (2, 0.051571062312939675),
 (3, 0.034179296351233165),
 (4, 0.31180478223116176),
 (5, 0.12274328238644315),
 (6, 0.024573659359149527),
 (7, 0.12487810821089254),
 (8, 0.03984095364447979),
 (9, 0.10832228060328025),
 (10, 0.1),
 (11, 0.06804138174397718),
 (12, 0.05735393346764045),
 (13, 0.026518576139191005),
 (14, 0.14569287935358965),
 (15, 0.05594542388644594),
 (16, 0.09622504486493764),
 (17, 0.11518245674418136),
 (18, 0.1301448015738384),
 (19, 0.06804138174397717),
 (20, 0.058925565098878974),
 (21, 0.07106690545187015),
 (22, 0.04233337566673017),
 (23, 0.05892556509887897),
 (24, 0.047673129462279626),
 (25, 0.055679425398421745),
 (26, 0.14852213144650117),
 (27, 0.18931004635463397),
 (28, 0.05634361698190111),
 (29, 0.041380294430118404),
 (30, 0.07671455149152696),
 (31, 0.14433756729740646),
 (32, 0.06759225880448438),
 (33, 0.06565321642986129),
 (34, 0.0),
 (35, 0.1336306209562122),
 (36, 0.14852213144650114),
 (

Sort the similarity array to get top 6 similiar Movie indexs

In [56]:
sorted(list(enumerate(similarity[0])),reverse=True,key=lambda x:x[1])[:6]

[(0, 1.0000000000000002),
 (2409, 0.44876373392787533),
 (4336, 0.43058755641446156),
 (373, 0.40147753427348315),
 (838, 0.3969283100678678),
 (1537, 0.3931785497463925)]

Created a Recommend function that will recommend movie based on Movie title

In [57]:
def recommend(movie,new_df):
    movie_index=new_df[new_df['title']==movie].index[0]
    distance=similarity[movie_index]
    movies_list=sorted(list(enumerate(distance)),reverse=True,key=lambda x:x[1])[1:6]
    
    for i in movies_list:
        # print(new_df.iloc[i[0]].movie_id,)
        print(new_df.iloc[i[0]].title)
recommend('Avatar',new_df)

Aliens
Silent Running
Mission to Mars
Alien³
Moonraker


In [58]:
recommend('Superman',new_df)

Superman II
Superman Returns
Man of Steel
Superman III
Superman IV: The Quest for Peace


In [59]:
new_df.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                tags  
0  In the 22nd century, a paraplegic Marine be di...  
1  Captain Barbossa, long believe to be dead, hav...  
2  A cryptic message from Bond’s past send him on...  
3  Following the death of District Attorney Harve...  
4  John Carter be a war-weary, former military ca...

Saving Our New df for Later

In [60]:
import pickle
pickle.dump(new_df,open('movies.pkl','wb'))

In [61]:
new_df.to_excel('movies.xlsx',index=False)

Save the Similarity array for Later use to calculater similarity.

In [62]:
pickle.dump(similarity,open('similarity.pkl','wb'))